packages/vite/src/node/build.ts

```diff
@@ -542,55 +544,65 @@ async function doBuild(
     parallelBuilds.push(bundle)
 
     const generate = (output: OutputOptions = {}) => {
-      return bundle[options.write ? 'write' : 'generate'](
-        buildOutputOptions(output)
-      )
+      return bundle[options.write ? 'write' : 'generate'](output)
     }
 
     if (options.write) {
-      prepareOutDir(outDir, options.emptyOutDir, config)
+      prepareOutDir(outDirs, options.emptyOutDir, config)
     }
 
-    if (Array.isArray(outputs)) {
-      const res = []
-      for (const output of outputs) {
-        res.push(await generate(output))
-      }
-      return res
-    } else {
-      return await generate(outputs)
+    const res = []
+    for (const output of normalizedOutputs) {
+      res.push(await generate(output))
     }
+    return Array.isArray(outputs) ? res : res[0]
   } catch (e) {
     outputBuildError(e)
     throw e
   }
 }
 
 function prepareOutDir(
-  outDir: string,
+  outDirs: string[],
   emptyOutDir: boolean | null,
   config: ResolvedConfig
 ) {
-  if (fs.existsSync(outDir)) {
-    if (
-      emptyOutDir == null &&
-      !normalizePath(outDir).startsWith(config.root + '/')
-    ) {
-      // warn if outDir is outside of root
-      config.logger.warn(
-        colors.yellow(
-          `\n${colors.bold(`(!)`)} outDir ${colors.white(
-            colors.dim(outDir)
-          )} is not inside project root and will not be emptied.\n` +
-            `Use --emptyOutDir to override.\n`
+  for (const outDir of new Set(outDirs)) {
+    if (fs.existsSync(outDir)) {
+      if (
+        emptyOutDir == null &&
+        !normalizePath(outDir).startsWith(config.root + '/')
+      ) {
+        // warn if outDir is outside of root
+        config.logger.warn(
+          colors.yellow(
+            `\n${colors.bold(`(!)`)} outDir ${colors.white(
+              colors.dim(outDir)
+            )} is not inside project root and will not be emptied.\n` +
+              `Use --emptyOutDir to override.\n`
+          )
         )
-      )
-    } else if (emptyOutDir !== false) {
-      emptyDir(outDir, ['.git'])
+      } else if (emptyOutDir !== false) {
+        // skip those other outDirs which are nested in current outDir
+        const skipDirs = outDirs
+          .map((dir) => {
+            const relative = path.relative(outDir, dir)
+            if (
+              relative &&
+              !relative.startsWith('..') &&
+              !path.isAbsolute(relative)
+            ) {
+              return relative
+            }
+            return ''
+          })
+          .filter(Boolean)
+        emptyDir(outDir, [...skipDirs, '.git'])
+      }
+    }
+    if (config.publicDir && fs.existsSync(config.publicDir)) {
+      copyDir(config.publicDir, outDir)
     }
```

Comment:
We are now copying public directory to every outDirs.
Is this what people expect?

IMO public files should be copied to `build.outDir`.
(though I'm not sure if there's a case when both public directory and `output.dir` is used)

Reply:
In partial of my cases, there are some external patch files in public dir. I would run vite in multiple threads to bundle them into different dirs via in the past.

So I think this will be more friendly to me (though I think this may a rare case).

Reply:
I see.
I'm not sure whether this change is safe to do in a minor/patch because it's slightly documented [here](https://vitejs.dev/config/shared-options.html#publicdir).
@patak-dev What do you think about this?

Reply:
Yes, I think it may be an unintended doc artifact though. If we support multiple `output.dir` then we should be copying the public files to each of them or the public feature would be broken.
But I agree that we should discuss a bit more about this change. Maybe we should document that multiple `output.dir` isn't supported and that you should instead run Vite multiple times? It seems wasteful though if the only diff is the format for example.
I've added it to the team meeting board.